step 4 bowling

youngsuseo opened the pull request:
리뷰 요청이 계속 늦어지게 되어 죄송합니다. :cry:
좋은 리뷰 항상 감사드립니다.
추가적으로 여쭤볼 내용들 코드 내부에 추가해두었는데 확인 부탁드리겠습니다. :pray:
감사합니다!

* Frame의 구조를 Frame(interface) <- AbstractFrame(abstract class, 공통 로직 추가) <- NormalFrame, FinalFrame으로 구현
* Frames 의 일급컬렉션, FrameLinkedList를 통해 LinkedList 형태로 구현

* State pattern 을 적용하여, State(interface) <- AbstractState(abstract class)
<- Finished, Running 형태로 구현하고 하위에 Strike, Spare, Miss, Gutter, Ready, FirstBowl 의 상태를 가지도록 구현
* StateEnum 을 상태의 인스턴스 타입을 확인
* FullFrameState : Frame이 가지는 상태를 묶어 클래스 분리

youngsuseo committed:
feat(bowling) : develop bowling game which is multiple players can play
* Frame의 구조를 Frame(interface) <- AbstractFrame(abstract class, 공통 로직 추가)
<- NormalFrame, FinalFrame으로 구현
* Frames 의 일급컬렉션, FrameLinkedList를 통해 LinkedList 형태로 구현

* State pattern 을 적용하여, State(interface) <- AbstractState(abstract class)
<- Finished, Running 형태로 구현하고 하위에 Strike, Spare, Miss, Gutter, Ready, FirstBowl 의 상태를 가지도록 구현
* StateEnum 을 상태의 인스턴스 타입을 확인
* FullFrameState : Frame이 가지는 상태를 묶어 클래스 분리

youngsuseo commented on src/main/java/bowling/BowlingMain.java:
```diff
@@ -1,32 +1,50 @@
+        for (int frameIndex = INITIAL_INDEX; frameIndex < NUMBERS_OF_FRAMES; frameIndex++) {
+            getPlayersScore(players, frameIndex);
+        }
+    }
```
Players를 만드는 과정을 factory 클래스로 이동해서 진행하려고 했었는데,
scanner를 통해 입력을 받는 부분이 factory 클래스에 섞이게 되는 것 같아 Main 부분에서 players를 만들도록 하였습니다.
이렇게 view를 호출하는 부분이 도메인이나 factory 클래스로 섞이게 된다면
따로 분리하지 않고 지금처럼 구현하는 것이 좋을까요?

ghojeong replied:
createPlayers private 메서드로 분리가 되어 있기 때문에, 충분히 가독성이 좋다고 생각합니다.
네 지금과 같은 모습 보기 좋습니다. 👍

youngsuseo commented on src/main/java/bowling/domain/frame/FinalFrame.java:
```diff
@@ -0,0 +1,53 @@
+
+        return fallenPinsCountOfFirstFrame() + fallenPinsCountOfSecondFrame() + bonusState.getFallenPins();
+    }
+
```
normal frame 에서는 현재 상태에 따라 다음 상태를 변경하는 것을 다형성을 통해 할 수 있었는데,
final frame 에서는 그게 조금 어려웠습니다. :cry:

4단계의 힌트를 보니 각 frame은 하나의 상태만 가지게 한다고 적혀있어서 적용하려고 했는데,
그렇게 구조를 잡았더니 normal frame은 구현할 수 있지만 final frame의 인스턴스 변수 설정과
gutter 상태의 처리가 어려워 기존 구조에 최대한 정리하는 형태로 구현했습니다..ㅠㅠ
혹시 normal frame 과 final frame의 인스턴스 변수를 어떻게 설정하면 좋을지 알려주실 수 있으실까요? :pray:

ghojeong replied:
기존 구조라 하면, AbstractFrame 에 있는 로직을 최대한 활용하는 형태를 취하셨다고 이해했습니다.
저는 충분히 예쁘게 다듬어졌다고 생각합니다. 

> normal frame 과 final frame의 인스턴스 변수를 어떻게 설정하면 좋을지 알려주실 수 있으실까요? 

혹시 어떤 방향으로 더 개선을 하고 싶어서 이 부분을 질문주신것인지 여쭈어봐도 될까요?
저는 모든 클래스의 멤버가 적절하게 잘 분배되었다고 생각이 되어서요.

youngsuseo replied:
제가 구현한 방법은 normal frame이 fullFrameState를 가지면서 내부에 firstHalfState, secondHalfState를 가지는 형태로 되어
하나의 normal frame안에 firstHalfFrame, secondHalfFrame의 두개의 상태를 가진다고 생각했습니다.
그러다보니 bowl 메서드를 수행할 때 이게 firstHalfFrame 인지 secondHalfFrame 인지 구별해야 하는 상황이 있었습니다.

강의 노트 힌트에서 기재되어있는 것 처럼 하나의 상태만 가지게 되면 firstHalfFrame인지, secondHalfFrame인지 구별할 필요가 없이
해당 state를 사용하면 되어서 normal frame 에서는 더 편해질 수 있을 것 같습니다.
다만 그렇게 되면 finalFrame의 인스턴스 변수는 기본 normal frame에 있는 state를 사용하고,
상태를 추가로 2개 더 가져야 하는 상황이 생기므로 인스턴스 변수를 두개 추가로 생성하면
state 상태를 관리하는 로직이 normal frame과 달라져서 상태를 관리하는 로직이 일관성이 없어 질 것이라 예상했습니다.

그런데 제가 이렇게 정리하다보니 normal frame은 하나의 state만 가지고
finalFrame 에서 추가로 state를 두개 더 가져도 큰 문제는 없을 것 같다고 생각이 들었습니다.
제가 좀 더 확인해봤어야 했는데ㅠㅠ 제가 질문하다가 정리가 되었네요..ㅎㅎ
지금 인스턴스 변수를 사용한 방식과 구현의 차이이며 비슷한 상황일 것 같습니다.
꼼꼼하게 확인해주셔서 감사합니다!

ghojeong commented on src/main/java/bowling/domain/frame/NormalFrame.java:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        return beforeBowl.getScore();
+    }
```
저는 오히려 FinalFrame 보다도 NormalFrame 의 연속된 if 문이 아쉽군요.
볼링 미션에서 추구하는 제일 중요한 부분은 연속된 if 문을 제거하는 리팩토링을 해보는 것이라서요.

youngsuseo replied:
네 말씀해주신 것을 듣고 추가적으로 점수를 계산하는 로직을 State에서 할 수 있도록 변경하였습니다.
꼼꼼하게 리뷰해주셔서 감사합니다 🙇

ghojeong commented on src/main/java/bowling/domain/state/Running.java:
```diff
@@ -0,0 +1,39 @@
+
+public abstract class Running extends AbstractState {
+
+    Running() {
+
+    }
+
+    Running(int fallenPins, String symbol) {
+        super(fallenPins, symbol);
+    }
+
+    @Override
+    public Score getScore() {
+        return new Score();
+    }
+
+    AbstractState bowlSecond(int secondBowl) {
```
Running class 에서 접근제어자를 명시하지 않은 메서드가 있는 것은 의도한 것인가요?

youngsuseo replied:
넵 제가 의도한건 해당 메서드를 패키지 내부의 하위 클래스에서만 사용하고 있어 default 접근제어자를 사용하였습니다.

ghojeong commented on src/main/java/bowling/domain/state/StateEnum.java:
```diff
@@ -0,0 +1,66 @@
+                                       .filter(stateEnum -> stateEnum.className.equals(state.getClass().getSimpleName()))
+                                       .findFirst().orElse(READY);
+        return superClassName.equals(matchedState.superClassName);
+    }
```
StateEnum 에서, 함수형 프로그래밍을 통해 로직을 처리하려 하신 점은 대단하신것 같습니다. 👍 
다만 boolean 타입의 리턴과, value 및 valueOfSuperClass 라는 메서드명은 처음 보는 사람이 보기에 직관적이지 않는 네이밍 같습니다.
메서드명만 좀더 자세하게 적어주시면 좋을 것 같아서 언급을 해봅니다.

youngsuseo replied:
네 해당 메서드명 변경하였습니다! 조언 감사합니다 🙂

ghojeong commented on src/main/java/bowling/BowlingMain.java:
```diff
@@ -1,32 +1,50 @@
+    private static final int NUMBERS_OF_FRAMES = 10;
     private static final int INITIAL_INDEX = 0;
 
     public static void main(String[] args) {
-        Name name = new Name(InputView.inputPlayerName());
+        Players players = createPlayers();
+        for (int frameIndex = INITIAL_INDEX; frameIndex < NUMBERS_OF_FRAMES; frameIndex++) {
+            getPlayersScore(players, frameIndex);
+        }
+    }
 
-        Frames frames = new Frames();
-        int frameIndex = INITIAL_INDEX;
+    private static Players createPlayers() {
+        int numberOfPlayers = InputView.inputNumberOfPlayers();
+
+        Map<Player, Frames> playerMap = new LinkedHashMap<>();
+        for (int i = 0; i < numberOfPlayers; i++) {
+            playerMap.put(new Player(InputView.inputPlayerName()), new Frames());
+        }
+        return new Players(playerMap);
+    }
```
BowlingMain 은 충분히 깔끔한것 같습니다.
다만 getPlayersScore 가 void 타입인데, getter 형태로 메서드 명을 지은 것은,
처음보는 사람이 보기에 오해를 할만한 여지가 있다고 생각합니다.
차라리 printPlayersScore 같은 것은 어떠신가요?

youngsuseo replied:
네! 말씀해주신 것으로 변경하면 더 가독성이 올라갈 것 같습니다. 
해당 메서드 이름으로 변경 완료하였습니다!

ghojeong reviewed (changes requested):
서영수님!

볼링미션의 마지막 Step 매우 훌륭하게 진행해주셨습니다.
얼마나 많은 시간과 노력이 들어갔는지, 코드만 읽어보아도 그 작업량에 감탄하게 되네요. 👍 
특히 테스트 코드와 병행하며 리팩토링을 진행해주신 점이 대단한 것 같습니다.

객체의 역할 및 상태 분배는 매우 적절하게 된 것 같아서, 지금의 구조를 변경할 필요는 없다고 생각합니다.
다만 처음 보는 사람이 읽기에 오해할 수 있는 메서드명이 몇개 눈에 띄어서,
메서드 명만 수정해주시면 승인을 해드리려 합니다.

마지막 Step 이면 힘이 빠질만도 한데, 마지막까지 엄청난 퀄리티로 작업을 해주셔서 감사합니다. 💯

youngsuseo committed:
Refactor(bowling) : refactor bowling game which is multiple players can play

* AbstractFrame, NormalFrame : Score의 left 인스턴스 변수를 조건으로 비교하는 로직을
State 의 calculateAdditionalScore 메서드로 이동
* Finished, Miss, Running, State : calculateAdditionalScore 구현
* IllegalCalculateException : calculateAdditionalScore이 동작해야하지 않는 경우 예외
* StateEnum : value -> compareStates, valueOfSuperClass -> compareSuperClassState 으로 메서드 이름 변경경
* Score : 점수를 콘솔창에 출력할 때 상태가 strike, spare 이면 추가 투구 이후에 점수를 계산하는데,
이처럼 점수를 아직 계산하지 않는 상황일 때 사용할 incompleteCalculation 정적 팩토리 메소드 추가
* BowlingMain : getPlayersScore -> printPlayersScore 메서드 이름 변

youngsuseo commented:
* AbstractFrame, NormalFrame : Score의 left 인스턴스 변수를 조건으로 비교하는 로직을
State 의 calculateAdditionalScore 메서드로 이동
* Finished, Miss, Running, State : calculateAdditionalScore 구현
* IllegalCalculateException : calculateAdditionalScore이 동작해야하지 않는 경우 예외
* StateEnum : value -> compareStates, valueOfSuperClass -> compareSuperClassState 으로 메서드 이름 변경경
* Score : 점수를 콘솔창에 출력할 때 상태가 strike, spare 이면 추가 투구 이후에 점수를 계산하는데,
이처럼 점수를 아직 계산하지 않는 상황일 때 사용할 incompleteCalculation 정적 팩토리 메소드 추가
* BowlingMain : getPlayersScore -> printPlayersScore 메서드 이름 변경

해당 내용으로 말씀해주신 부분들 수정하였습니다!

ghojeong reviewed (approved):
서영수님!
마지막 미션까지 훌륭하게 따라와주셔서 정말 감사합니다.
공들인 코드는 광이 난다고 저는 표현하는데, ✨ 
영수님께서 작성하신 코드에 반짝반짝 빛이 나는 것 같습니다. 

열심히 해주셔서 다시한번 감사드리며,
많은 걸 배워가셨기를 바랍니다.
